### PR TITLE
Migrate scripted tests for sbt bridge from zinc repo [ci: last-only]

### PR DESCRIPTION
**Pull request by lrytz**
Build changes copied and adapted from lampepfl/dotty. Tests copied and adapted from sbt/zinc.

To Do
  - [x] `source-dependencies/patMat-scope`: after first change, compilation only seems to compile changed file. Existing ticket: https://github.com/sbt/zinc/issues/1229
  - [x] `source-dependencies/empty-package`: after setting `withRecompileAllFraction` to 1.0 this started failing. Old issue, logged at https://github.com/sbt/zinc/issues/1268
  - [x] Implement missing check operations: `checkMainClasses`, `checkIterations`, `checkWarnings`, ~`checkRecompilations`~, ~`checkDependencies`~, `checkClasses`, ~`checkProducts`~, `checkNoClassFiles`, ~`checkProductsExists`~, `checkNameExistsInClass`, re-enable tests that were disabled in the corresponding commit (one of this PR)
  - [x] check defaults that are different in zinc's scripted implementation (https://github.com/sbt/zinc/blob/9431f67d02feac2456aea50d194e903008f7de79/internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala#L294)
  - [x] ~figure out when to run the scripted tests in CI~ [done](https://github.com/scala/scala/pull/10554#issuecomment-1805954487)
    - On my machine running them all takes about 15 minutes
    - In the zinc repo tests run much faster; zinc is not using sbt's implementation for `scripted` tests but rather has its own re-implementation of the a "scripted" test framework.
    - Setting `scriptedBatchExecution := true` would make them run much faster, however there's cross-talk between tests that makes some of them fail (e.g., recompilation counts don't start at 1). In dotty some such issues were adressed in https://github.com/lampepfl/dotty/pull/10498/commits, but I'd prefer to keep that complexity out.
    - The fact that each test performs a reload (`> setup; reload` to include the shared settings) doesn't seem to affect performance too much (tested by copying the shared sbt file to each test directory).
  - [x] Check if there are tests that seem irrelevant to testing the compiler bridge and could be deleted.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -943,6 +943,50 @@ def osgiTestProject(p: Project, framework: ModuleID) = p
     cleanFiles += (ThisBuild / buildDirectory).value / "osgi"
   )
 
+lazy val sbtTest = project.in(file("test") / "sbt-test")
+  .enablePlugins(ScriptedPlugin)
+  .settings(disableDocs)
+  .settings(
+    scalaVersion := "2.12.18",
```
Comment (lrytz) [inline build.sbt]: This should be the Scala version on which sbt is running. yoshord on discord suggested

> add sbt-buildinfo to the meta-metabuild, and generate a BuildInfo.scala for the metabuild to use

**Comment by som-snytt** [comment]
Some commits are like, "I wish I'd taken that class in college."

**Review by dwijnand — APPROVED** [review]

**Comment by lrytz** [comment]
This is done now, modulo deciding how to run the tests.

@SethTisue see my comments in the PR description. What do you think?

**Comment by lrytz** [comment]
Managed to get `scriptedBatchExecution := true` working, which also opened up the door for `scriptedParallelInstances := 2`. Now the whole suite runs in 2:11 on my computer 🏁 so we can enable it in CI (already done in this PR). Thanks @dwijnand for the help!

**Review by SethTisue — APPROVED** [review]
LGTM, but maybe squash into fewer commits before merging?

**Comment by lrytz** [comment]
👍  will do

**Pull request merged**